Reduced resources for various Akka dispatchers

hubertp opened the pull request:
### Pull Request Description

The defaults picked up by Akka tend to make us of all resources which is unnecessary and overwhelming for tests.

Improves #8806, potentially.

Before
![Screenshot from 2024-01-28 22-34-42](https://github.com/enso-org/enso/assets/292128/f80eb66a-2f37-44d5-bcdb-f00a78fe72fd)
After
![Screenshot from 2024-01-31 00-12-10](https://github.com/enso-org/enso/assets/292128/c5223912-5f6e-413c-a0a4-050afa3ed463)

when running the problematic `LibrariesTest`.

Full `language-server` test suite.
Before
![Screenshot from 2024-01-31 00-20-50](https://github.com/enso-org/enso/assets/292128/f1c94a66-6905-4f57-8a7d-7df049714353)
After
![Screenshot from 2024-01-31 00-18-40](https://github.com/enso-org/enso/assets/292128/3a11125e-d593-43df-8d35-1a8915812b2b)


### Important Notes

Note that Executors assigned to Zio and initializers should also be improved. Unfortunately due to various blocking threadpools  it is easy to get timeouts when running the whole suite.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

hubertp committed:
Reduced resources for various Akka dispatchers

The defaults picked up by Akka tend to make us of all resources which is
unnecessary and overwhelming for tests.

Note that Executors assigned to Zio and initializers should also be
improved but there are timeouts when running the whole suite. TBD

hubertp committed:
No more multiple search.db threads in tests

We weren't shutting down sql database.

JaroslavTulach commented:
> defaults picked up by Akka tend to make us(e) of all resources which is unnecessary and overwhelming for tests.

That's unnecessary for an application/web server that handles single client 95% of times as well!

JaroslavTulach commented on engine/language-server/src/test/scala/org/enso/languageserver/boot/resource/RepoInitializationSpec.scala:
```diff
@@ -35,7 +35,8 @@ class RepoInitializationSpec
 
   override def afterAll(): Unit = {
-    TestKit.shutdownActorSystem(system)
+    TestKit.shutdownActorSystem(system, verifySystemShutdown = true)
```
There is a `TestKit` support in Akka and calling `shutdownActorSystem` doesn't really shut it down, unless one `verifySystemShutdown`!? Good you find it.

hubertp replied:
I _think_ it was mostly shutting it down. This ensures that it really does it so that I don't have to chase ghosts.

JaroslavTulach commented on engine/language-server/src/test/resources/application-test.conf:
```diff
@@ -4,6 +4,63 @@ akka.http.server.websocket.periodic-keep-alive-max-idle = 1 second
+  executor = "thread-pool-executor"
+  # Configuration for the thread pool
+  thread-pool-executor {
+    fixed-pool-size = 2
```
Have you tried `fixed-pool-size = 1`? Does the application work? If so, _single threaded executor_ is something I tend to prefer...

hubertp replied:
Yes. We are talking about 70s execution time vs 800s, on my machine. I think it is not worth it.

hubertp commented on engine/language-server/src/test/scala/org/enso/languageserver/websocket/json/BaseServerTest.scala:
```diff
@@ -179,6 +179,11 @@ class BaseServerTest
   }
 
+  override def afterAll(): Unit = {
+    sqlDatabase.close()
```
That's why you were seeing so many `searcher.db` threads, @JaroslavTulach

hubertp committed:
make sure to close actor systems

hubertp committed:
Eliminate spurious test suits

Any class extending `Suite` is treated as a potential test suite to run.
Even if it is class that abstracts away some common functionality.
Since `BaseBinaryServerTest` and `BaseServerTest` start full Actor's
system, the resources were pretty wasted at that point. Those base
classes need to be abstract to not be treated as suites.

hubertp commented:
I've managed to cut threads by half when running full `language-server` suite. 
There are still areas to improve as indicated by the following screenshots:
![Screenshot from 2024-01-31 00-23-51](https://github.com/enso-org/enso/assets/292128/32858e53-0047-4b8a-aff6-cf49bff594d6)
![Screenshot from 2024-01-31 00-24-11](https://github.com/enso-org/enso/assets/292128/e358691f-bac3-4235-bb6a-7fd4c4e97092)
1. The whole ZIO stuff is a real PITA to configure, at least in the current setting. But it is also the most expensive so probably worth looking into.
2. `http-requests` threads are from our implementation of http downloader which a) uses Akka (one unicorn dies) b) does not terminate Akka system until the whole process dies (another unicorn dies). I'm very much in favour of just ditching akka in that library and replacing the logic with something simpler.
3. Some random threads that are not being shutdown

So far, I think this PR improves the state.

hubertp committed:
Merge branch 'develop' into wip/hubert/8806-reduce-resources

hubertp committed:
One more thread pool to close

hubertp committed:
Enable flaky project now

Akirathan commented:
> http-requests threads are from our implementation of http downloader which a) uses Akka (one unicorn dies) b) does not terminate Akka system until the whole process dies (another unicorn dies). I'm very much in favour of just ditching akka in that library and replacing the logic with something simpler.

@hubertp FYI: That is tracked in https://github.com/enso-org/enso/issues/8222
